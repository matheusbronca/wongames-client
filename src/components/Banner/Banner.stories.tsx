import { Meta, StoryObj } from '@storybook/react'
import Banner, { BannerProps } from './Banner'

const commonProps: BannerProps = {
  title: 'Super Mario World',
  subtitle:
    "At <strong>Super Mario World</strong> you're going to pass through a lot of challenges, are you ready for this adventure?",
  headingTag: 'h2',
  img: 'https://images.unsplash.com/photo-1629646526325-d4e092b641b3?q=80&w=1974&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D',
  buttonLabel: 'Buy now',
  buttonLink: '/#'
}

export default {
  title: 'Banner',
  args: commonProps,
  component: Banner
} as Meta

export const Mobile: StoryObj<BannerProps> = {
  render: () => (
    <div className="bg-mainBg size-full absolute">
      <Banner {...commonProps} />
    </div>
  )
}

Mobile.parameters = {
  viewport: {
    defaultViewport: 'mobile1'
  },
  layout: 'fullscreen'
} as Meta

export const Desktop: StoryObj<BannerProps> = {
  render: () => (
    <div className="bg-gradient-to-b from-mainBg via-mainBg via-50% to-white to-50% size-full min-h-full absolute">
      <div className="container my-auto size-full grid grid-cols-1 place-content-center">
        <Banner {...commonProps} />
      </div>
    </div>
  )
}

Desktop.parameters = {
  viewport: {
    defaultViewport: 'desktop'
  },
  layout: 'fullscreen'
}

export const DesktopWithRibbon: StoryObj = {
  render: () => (
    <div className="bg-gradient-to-b from-mainBg via-mainBg via-50% to-white to-50% size-full min-h-full absolute">
      <div className="container my-auto size-full grid grid-cols-1 place-content-center">
        <Banner {...commonProps} ribbon="My Ribbon" />
      </div>
    </div>
  )
}

DesktopWithRibbon.parameters = {
  viewport: {
    defaultViewport: 'desktop'
  },
  layout: 'fullscreen'
}

export const MobileWithRibbon: StoryObj<BannerProps> = {
  render: () => (
    <div className="bg-mainBg size-full absolute">
      <Banner {...commonProps} ribbon="My Ribbon" />
    </div>
  )
}

MobileWithRibbon.parameters = {
  viewport: {
    defaultViewport: 'mobile1'
  },
  layout: 'fullscreen'
}
