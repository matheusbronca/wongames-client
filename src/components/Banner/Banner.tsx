import React from 'react'
import Button from '../Button/Button'
import MatchMedia from '../MatchMedia/MatchMedia'
import Ribbon from '../Ribbon/Ribbon'
import { RibbonColor, RibbonSize } from '../Ribbon/Ribbon'

/* eslint-disable @next/next/no-img-element */

// type BannerType<T> = (props: BannerProps<T>) =>
export type HType = Extract<
  React.ElementType,
  'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'
>

export type BannerProps = {
  img: string
  title: string
  subtitle: string
  headingTag?: HType
  buttonLabel: string
  buttonLink: string
  ribbon?: string
  ribbonSize?: RibbonSize
  ribbonColor?: RibbonColor
}

const Banner = ({
  img,
  title,
  headingTag = 'h3',
  subtitle,
  buttonLabel,
  buttonLink,
  ribbon,
  ribbonSize = 'normal',
  ribbonColor = 'primary'
}: BannerProps) => {
  const HeadingTag = ({
    as,
    children,
    ...props
  }: {
    as: HType
    children: string
  } & React.HTMLProps<HTMLHeadingElement>) =>
    React.createElement(as, props, children)

  return (
    <article className="relative w-full md:relative md:rounded-md md:shadow-xl ">
      {!!ribbon && (
        <Ribbon
          color={ribbonColor}
          size={ribbonSize}
          className="right-0 md:-right-4 after:hidden md:after:block"
        >
          {ribbon}
        </Ribbon>
      )}
      <figure className="w-full h-60 md:h-[36rem]">
        <img
          src={img}
          alt={`${title} - ${subtitle}`}
          className="size-full object-cover md:rounded-md"
        />
        <figcaption className="text-white p-8 bg-opacity-15 bg-white md:absolute md:bottom-0 md:left-0 md:right-0 md:bg-[#000000] md:bg-opacity-70 md:p-11 md:rounded-md md:rounded-t-none">
          <HeadingTag
            as={headingTag}
            className="text-xl md:text-3xl font-semibold"
          >
            {title}
          </HeadingTag>
          <p
            className="text-sm mt-xxsm mb-xsm [&_>_strong]:text-primary [&_>_strong]:font-semibold md:text-lg"
            dangerouslySetInnerHTML={{ __html: subtitle }}
          />
          <MatchMedia matcher="lessThan: md">
            <Button role="a" href={buttonLink}>
              {buttonLabel}
            </Button>
          </MatchMedia>
          <MatchMedia matcher="greaterThan: md">
            <Button role="a" href={buttonLink} size="large">
              {buttonLabel}
            </Button>
          </MatchMedia>
        </figcaption>
      </figure>
    </article>
  )
}

export default Banner
