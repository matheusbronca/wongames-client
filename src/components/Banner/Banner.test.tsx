import { render, screen } from '@testing-library/react'

import Banner from './Banner'

const props = {
  title: 'Banner',
  subtitle:
    "At Super Mario World you're going to pass through a lot of challenges, are you ready for this adventure?",
  img: 'https://images.unsplash.com/photo-1629646526325-d4e092b641b3?q=80&w=1974&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D',
  buttonLabel: 'Buy now',
  buttonLink: '/#'
}

describe('<Banner />', () => {
  it('shoudl render heading as h3 by default', () => {
    render(<Banner {...props} />)

    expect(
      screen.getByRole('heading', { name: /Banner/i, level: 3 })
    ).toBeInTheDocument()
  })

  it('should render correctly', () => {
    const { container } = render(<Banner {...props} headingTag="h2" />)

    // Se o titulo existe
    expect(
      screen.getByRole('heading', { name: /Banner/i, level: 2 })
    ).toBeInTheDocument()
    //
    // Se o subtitulo existe
    expect(screen.getByRole('paragraph')).toHaveTextContent(
      /At Super Mario World you're going to pass through a lot of challenges, are you ready for this adventure\?/i
    )
    // se a imagem existe
    expect(screen.getByRole('img')).toBeInTheDocument()
    expect(
      screen.getByRole('img', {
        name: /Banner - At Super Mario World you're going to pass through a lot of challenges, are you ready for this adventure?/i
      })
    ).toHaveAttribute('src')

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a Ribbon', () => {
    render(
      <Banner
        {...props}
        ribbon="My Ribbon"
        ribbonSize="small"
        ribbonColor="secondary"
      />
    )

    const ribbon = screen.getByText(/My Ribbon/i)

    expect(ribbon).toBeInTheDocument()
    expect(ribbon).toHaveClass('bg-secondary after:border-secondary')
    expect(ribbon).toHaveClass('after:size-[6px] text-xs')
  })
})
