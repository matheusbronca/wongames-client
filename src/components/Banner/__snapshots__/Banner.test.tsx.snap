// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should render correctly 1`] = `
<article
  class="relative w-full md:relative md:rounded-md md:shadow-xl "
>
  <figure
    class="w-full h-60 md:h-[36rem]"
  >
    <img
      alt="Banner - At Super Mario World you're going to pass through a lot of challenges, are you ready for this adventure?"
      class="size-full object-cover md:rounded-md"
      src="https://images.unsplash.com/photo-1629646526325-d4e092b641b3?q=80&w=1974&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D"
    />
    <figcaption
      class="text-white p-8 bg-opacity-15 bg-white md:absolute md:bottom-0 md:left-0 md:right-0 md:bg-[#000000] md:bg-opacity-70 md:p-11 md:rounded-md md:rounded-t-none"
    >
      <h2
        class="text-xl md:text-3xl font-semibold"
      >
        Banner
      </h2>
      <p
        class="text-sm mt-xxsm mb-xsm [&_>_strong]:text-primary [&_>_strong]:font-semibold md:text-lg"
      >
        At Super Mario World you're going to pass through a lot of challenges, are you ready for this adventure?
      </p>
      <div
        class="block md:hidden"
      >
        <button
          class="inline-flex items-center justify-center text-white no-underline border-none rounded bg-gradient-to-b cursor-pointer from-[#FF5F5F] to-[#F062C0] hover:bg-gradient-to-b hover:from-[#e35565] hover:to-[#d958a6] text-sm h-10 px-md"
          href="/#"
          role="a"
        >
          <span>
            Buy now
          </span>
        </button>
      </div>
      <div
        class="hidden md:block"
      >
        <button
          class="inline-flex items-center justify-center text-white no-underline border-none rounded bg-gradient-to-b cursor-pointer from-[#FF5F5F] to-[#F062C0] hover:bg-gradient-to-b hover:from-[#e35565] hover:to-[#d958a6] text-base h-12 px-xlg"
          href="/#"
          role="a"
        >
          <span>
            Buy now
          </span>
        </button>
      </div>
    </figcaption>
  </figure>
</article>
`;
